### PR TITLE
[WIP] test: added test scripts for the functional tests

### PR DESCRIPTION
**Pull request by kayrus**
subj

### CODE REVIEW

**Comment by jonboulle** [comment]
I don't understand the use case for this. can you help me understand what it provides over just a bash script that runs this? Since we still need etcd, systemd etc running on the host.

**Comment by kayrus** [comment]
It is still WIP. And this script will be executed on coreos.

**Comment by tixxdz** [comment]
You may also check if etcd is up.

**Comment by jonboulle** [comment]
What is the point of doing this in docker at all? Look how much stuff you necessarily need to bring in from the host..

**Comment by kayrus** [comment]
@jonboulle this will allow you to run tests directly on coreos. which already has etcd.

**Comment by jonboulle** [comment]
Why do you need docker for that? The functional tests run fine on coreos
without it

kayrus <notifications@github.com> schrieb am Fr., 12. Feb. 2016 11:50:

**Comment by kayrus** [comment]
@jonboulle there is not go binary inside coreos.

**Comment by antrik** [comment]
To elaborate a bit: according to `functional/README.md`, the official way to run tests is inside a CoreOS VM, typically through Vagrant. However, as CoreOS doesn't provide a Go environment itself, the README suggest mapping it from outside the VM using Vagrant shared folders. But that's neither easy nor a self-contained reproducible environment. In our understanding, the canonical way to get a Go environment inside CoreOS is by running a golang container.

**Comment by jonboulle** [comment]
@antrik thanks for the context, very helpful. I discussed out of band with @mirkoboehm and we'll circulate notes shortly

**Comment by kayrus** [comment]
@jonboulle tests passed. should be squash and merge?

File: functional/platform/nspawn.go [hunk 1]
```diff
@@ -302,14 +302,18 @@ func (nc *nspawnCluster) createMember(id string) (m Member, err error) {
 
 		// minimum requirements for running systemd/coreos in a container
 		fmt.Sprintf("mkdir -p %s/usr", fsdir),
+		//fmt.Sprintf("mkdir -p %s/lib", fsdir), //ubuntu
```
Comment (jonboulle) [inline functional/platform/nspawn.go]: can you explain all these?

File: functional/platform/nspawn.go [hunk 2]
```diff
@@ -395,6 +401,7 @@ UseDNS no
 			return
 		default:
 		}
+		log.Printf("Dialing machine: %s", addr)
```
Comment (jonboulle) [inline functional/platform/nspawn.go]: do we need this?

File: functional/start_etcd [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash -e
```
Comment (jonboulle) [inline functional/start_etcd]: what is this script about?

File: functional/platform/nspawn.go (hunk 1 above)
Comment (kayrus) [inline functional/platform/nspawn.go]: I tried to run these tests in ubuntu without cloud-config. Decided to keep this. If you don't like - i'll remove.

File: functional/platform/nspawn.go (hunk 2 above)
Comment (kayrus) [inline functional/platform/nspawn.go]: Yes, otherwise we don't have any feedback from tests. btw, we already have `fleetctl list-machines... ` in output. Why don't print `Dialing`?

File: functional/start_etcd (hunk 3 above)
Comment (kayrus) [inline functional/start_etcd]: For those who want to run tests in qemu (and any other coreos environments) without specified cloud-config.

File: build [hunk 4]
```diff
@@ -1,29 +1,31 @@
 #!/bin/bash -e
 
+CDIR=$(cd `dirname $0` && pwd)
```
Comment (jonboulle) [inline build]: is this working around some problem you encountered?

File: functional/README.md [hunk 5]
```diff
@@ -1,3 +1,63 @@
...
+sudo apt-get update
+sudo apt-get install -y build-essential dkms
+sudo apt-get install -y VirtualBox-5.0
+#Previous VirtualBox (if you have problems with nested virtualization, more info here: https://www.virtualbox.org/ticket/14965)
+#sudo apt-get install -y VirtualBox-4.3
+```
+
+### Centos/Fedora
+
+**NOTE**: NFS doesn't work on CentOS 6.x
```
Comment (jonboulle) [inline functional/README.md]: what does this imply? that they must use >=7?

File: functional/README.md [hunk 6]
```diff
@@ -1,3 +1,63 @@
...
+git clone https://github.com/coreos/fleet
+cd fleet/functional
+# wget https://raw.githubusercontent.com/coreos/coreos-vagrant/master/Vagrantfile
+# we will use default unmodified CoreOS Vagrantfile
+./run-in-vagrant
+```
+
 ## fleet functional tests
 
 This functional test suite deploys a fleet cluster using nspawn containers, and asserts fleet is functioning properly.
```
Comment (jonboulle) [inline functional/README.md]: This intro should still be at the top of the file, otherwise the configure stuff lacks context

File: functional/Vagrantfile [hunk 7]
```diff
@@ -0,0 +1,145 @@
+# -*- mode: ruby -*-
```
Comment (jonboulle) [inline functional/Vagrantfile]: is this just a straight port of coreos-vagrant, plus  the provision script?

File: functional/provision/setup.sh [hunk 8]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash -e
+
+USER_ID=${SUDO_UID:-$(id -u)}
+HOME=$(getent passwd "${USER_ID}" | cut -d: -f6)
+
+export GOROOT=${HOME}/go
+export PATH=${HOME}/go/bin:${PATH}
+
+if [ ! -f ${HOME}/go1.5.3.linux-amd64.tar.gz ]; then
```
Comment (jonboulle) [inline functional/provision/setup.sh]: split out go1.5.3.linux-amd64.tar.gz as a var

File: functional/provision/setup.sh [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  # Remove unfinished archive when you press Ctrl+C
```
Comment (jonboulle) [inline functional/provision/setup.sh]: this seems unnecessary, why not just download it in a mktemp and extract it into HOME?

File: functional/start_etcd (hunk 3 above)
Comment (jonboulle) [inline functional/start_etcd]: OK, this needs to be documented, it's not at all self explanatory

File: build (hunk 4 above)
Comment (kayrus) [inline build]: Yes, when you run scripts from home directory for example. `cd ~ ; sudo ./fleet./functional/test`

File: functional/start_etcd [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+USER_ID=${SUDO_UID:-$(id -u)}
+HOME=$(getent passwd "${USER_ID}" | cut -d: -f6)
```
Comment (jonboulle) [inline functional/start_etcd]: I don't really understand all these machinations. Why do we need HOME? why do we need to ensure we're sudo and not the root user? can't you just put it in temp and run it from there?

File: functional/README.md (hunk 5 above)
Comment (kayrus) [inline functional/README.md]: Don't want to waste my time trying to debug why it doesn't work in old CentOS 6. So yes.

File: functional/start_etcd (hunk 3 above)
Comment (jonboulle) [inline functional/start_etcd]: e.g. maybe they should run coreos-cloudinit --from-file in that case?

File: functional/Vagrantfile (hunk 7 above)
Comment (reneengelhard) [inline functional/Vagrantfile]: yes

Comment (kayrus) [inline functional/Vagrantfile]: The only one string we added is `config.vm.provision :shell, :path => "provision/setup.sh", :privileged => false`. The rest is original coreos-vagrant Vagrantfile.

File: functional/README.md (hunk 5 above)
Comment (jonboulle) [inline functional/README.md]: That's fine, I'm just trying to suggest explicitness

File: functional/provision/setup.sh (hunk 9 above)
Comment (kayrus) [inline functional/provision/setup.sh]: Sounds reasonable.

File: functional/platform/nspawn.go (hunk 1 above)
Comment (jonboulle) [inline functional/platform/nspawn.go]: We definitely shouldn't leave commented lines in here. Are the lines totally superfluous now? The tests all work fine without?

File: functional/platform/nspawn.go [hunk 11]
```diff
@@ -302,14 +302,18 @@ func (nc *nspawnCluster) createMember(id string) (m Member, err error) {
...
+		fmt.Sprintf("echo 'core:x:500:500:CoreOS Admin:/home/core:/bin/bash' > %s/etc/passwd", fsdir),
+		fmt.Sprintf("echo 'core:x:500:' > %s/etc/group", fsdir),
 		fmt.Sprintf("ln -s /proc/self/mounts %s/etc/mtab", fsdir),
 		fmt.Sprintf("ln -s usr/lib64 %s/lib64", fsdir),
 		fmt.Sprintf("ln -s lib64 %s/lib", fsdir),
 		fmt.Sprintf("ln -s usr/bin %s/bin", fsdir),
 		fmt.Sprintf("ln -s usr/sbin %s/sbin", fsdir),
 		fmt.Sprintf("mkdir -p %s/home/core/.ssh", fsdir),
-		fmt.Sprintf("chown -R core:core %s/home/core", fsdir),
+		fmt.Sprintf("chown -R 500:500 %s/home/core", fsdir),
```
Comment (jonboulle) [inline functional/platform/nspawn.go]: is this necessary after you just added the user above?

File: functional/provision/setup.sh [hunk 12]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash -e
```
Comment (jonboulle) [inline functional/provision/setup.sh]: install_go.sh

File: functional/test [hunk 13]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash -e
+
+CDIR=$(cd `dirname $0` && pwd)
+USER_ID=${SUDO_UID:-$(id -u)}
+HOME=$(getent passwd "${USER_ID}" | cut -d: -f6)
+
+cd ${CDIR}/../
```
Comment (jonboulle) [inline functional/test]: If you're going to the trouble of changing to the root of the repo here, then just use relative paths for the rest of the script. Otherwise it's too much to maintain

File: functional/start_etcd (hunk 10 above)
Comment (kayrus) [inline functional/start_etcd]: Temp will be cleaned up after reboot. And after reboot IP address could be also changed. This script also works under root user.

File: functional/platform/nspawn.go (hunk 1 above)
Comment (kayrus) [inline functional/platform/nspawn.go]: Yes, they do. Will remove ubuntu stuff. But will keep `.. 500 ...` for future support.

File: functional/provision/setup.sh (hunk 12 above)
Comment (reneengelhard) [inline functional/provision/setup.sh]: ok, one might (for whatever reason) want to do other stuff in setup.sh, though, so I'll make setup.sh call a new install_go.sh

File: functional/platform/nspawn.go (hunk 11 above)
Comment (kayrus) [inline functional/platform/nspawn.go]: Yes, otherwise user directory will be owned by root

File: functional/provision/setup.sh (hunk 8 above)
Comment (kayrus) [inline functional/provision/setup.sh]: Will fix. And note above too.

File: functional/start_etcd (hunk 3 above)
Comment (kayrus) [inline functional/start_etcd]: after we run `setup-network-environment` - yes. will fix that.

File: functional/README.md (hunk 6 above)
Comment (kayrus) [inline functional/README.md]: True. Will fix.

File: functional/test (hunk 13 above)
Comment (kayrus) [inline functional/test]: Relatives paths will not help when you run scripts from the outside, i.e. `cd ~ ; sudo ./fleet/functional/test`

File: functional/provision/setup.sh (hunk 12 above)
Comment (reneengelhard) [inline functional/provision/setup.sh]: nevermind my last comment; did a complete move now anyways as what might make sense in a setup.sh might need privileges which we donÄ't need for install_go.sh and if we ever needed it we can run extra setup.sh

File: functional/platform/nspawn.go (hunk 11 above)
Comment (jonboulle) [inline functional/platform/nspawn.go]: I mean why can't it be core:core

File: functional/test (hunk 13 above)
Comment (kayrus) [inline functional/test]: Fixed

File: functional/platform/nspawn.go (hunk 11 above)
Comment (kayrus) [inline functional/platform/nspawn.go]: Because we run this command on host, not inside container and it will not work for non-coreos tests in future.

**Comment by kayrus** [comment]
@jonboulle done, please review once again.

File: functional/README.md [hunk 14]
```diff
@@ -4,50 +4,93 @@ This functional test suite deploys a fleet cluster using nspawn containers, and
...
+$ sudo fleet/functional/start_etcd
 ```
-$ systemctl start etcd
+
+It will configure and start one-member etcd2 daemon.
+
+Then run the functional tests:
+
+```sh
+$ git clone https://github.com/coreos/fleet
```
Comment (jonboulle) [inline functional/README.md]: haven't we already cloned?

File: functional/README.md [hunk 15]
```diff
@@ -4,50 +4,93 @@ This functional test suite deploys a fleet cluster using nspawn containers, and
...
+$ sudo fleet/functional/test
 ```
 
-Then the tests can be run with:
+## configure environment to run Vagrant
```
Comment (jonboulle) [inline functional/README.md]: Configure host environment to run Vagrant

File: functional/README.md [hunk 16]
```diff
@@ -4,50 +4,93 @@ This functional test suite deploys a fleet cluster using nspawn containers, and
...
-```
-2. Ensure the `FLEETD_BIN` and `FLEETCTL_BIN` environment variables point to the respective fleetd and fleetctl binaries that should be used to drive the actual tests.
+### run tests in vagrant
 
+Vagrant will configure CoreOS instance with one-member etcd2 daemon.
+
+```sh
+$ git clone https://github.com/coreos/fleet
+$ cd fleet/functional
+$ # Vagrantfile based on official CoreOS Vagranfile with one extra provision string
```
Comment (jonboulle) [inline functional/README.md]: can this comment instead go in the vagrantfile?

File: functional/README.md [hunk 17]
```diff
@@ -4,50 +4,93 @@ This functional test suite deploys a fleet cluster using nspawn containers, and
 
 It shares an instance of etcd deployed on the host machine with each of the nspawn containers.
 
-It's recommended to run this in a virtual machine environment on CoreOS (e.g. using coreos-vagrant). The only dependency for the tests not provided on the CoreOS image is `go`.
+It's recommended to run this in a virtual machine environment on CoreOS (e.g. using [Vagrant][test-in-vagrant]).
 
-The caller must do three things before running the tests:
+Since the tests utilize `[systemd-nspawn]`, this needs to be invoked as sudo/root.
```
Comment (jonboulle) [inline functional/README.md]: this needs -> they need

File: functional/README.md [hunk 18]
```diff
@@ -4,50 +4,93 @@ This functional test suite deploys a fleet cluster using nspawn containers, and
...
+If the tests are aborted partway through, it's currently possible for them to leave residual state as a result of the `systemd-nspawn` operations. This can be cleaned up using the `clean.sh` script.
 
-```
-$ ssh-agent
-$ ssh-add fleet/functional/fixtures/id_rsa
-$ echo $SSH_AUTH_SOCK
-/tmp/ssh-kwmtTOsL7978/agent.7978
-```
-2. Ensure the `FLEETD_BIN` and `FLEETCTL_BIN` environment variables point to the respective fleetd and fleetctl binaries that should be used to drive the actual tests.
+### run tests in vagrant
```
Comment (jonboulle) [inline functional/README.md]: Run tests in Vagrant

File: functional/README.md [hunk 19]
```diff
@@ -4,50 +4,93 @@ This functional test suite deploys a fleet cluster using nspawn containers, and
...
 
+Vagrant will configure CoreOS instance with one-member etcd2 daemon.
```
Comment (jonboulle) [inline functional/README.md]: The recommended way to run the tests is to use the provided Vagrantfile, which will set up a single CoreOS instance with a one-member etcd cluster.

File: functional/README.md [hunk 20]
```diff
@@ -4,50 +4,93 @@ This functional test suite deploys a fleet cluster using nspawn containers, and
...
+```sh
+$ git clone https://github.com/coreos/fleet
+$ cd fleet/functional
+$ # Vagrantfile based on official CoreOS Vagranfile with one extra provision string
+$ ./run-in-vagrant
 ```
-$ export FLEETD_BIN=/path/to/fleetd
-$ export FLEETCTL_BIN=/path/to/fleetctl
+
+### run tests inside other CoreOS platforms (QEMU/BareMetal/libvirt/etc)
```
Comment (jonboulle) [inline functional/README.md]: Run

File: functional/README.md [hunk 21]
```diff
@@ -4,50 +4,93 @@ This functional test suite deploys a fleet cluster using nspawn containers, and
...
+
```
Comment (jonboulle) [inline functional/README.md]: It's also possible to run the tests on CoreOS on other platforms

File: functional/README.md [hunk 22]
```diff
@@ -4,50 +4,93 @@ This functional test suite deploys a fleet cluster using nspawn containers, and
...
 
-3. Make sure etcd is running on the host system.
+If you didn't configure etcd2 daemon yet, just run this script:
 
+```sh
+$ sudo fleet/functional/start_etcd
 ```
-$ systemctl start etcd
+
+It will configure and start one-member etcd2 daemon.
```
Comment (jonboulle) [inline functional/README.md]: a one-member etcd cluster

File: functional/config.rb [hunk 23]
```diff
@@ -0,0 +1,58 @@
+# Size of the CoreOS cluster created by Vagrant
```
Comment (jonboulle) [inline functional/config.rb]: Do we need this configuration? It's just one more moving part. If we have already modified the Vagrantfile, then just do it in there and remove this file.

File: functional/platform/nspawn.go (hunk 11 above)
Comment (jonboulle) [inline functional/platform/nspawn.go]: Oh, I missed we don't use systemd-nspawn for this, sorry.

File: functional/platform/nspawn.go (hunk 2 above)
Comment (jonboulle) [inline functional/platform/nspawn.go]: I don't follow?

File: functional/test-docker [hunk 24]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash -e
```
Comment (jonboulle) [inline functional/test-docker]: Do we need this any more?

File: functional/Vagrantfile [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+      if $share_home
+        config.vm.synced_folder ENV['HOME'], ENV['HOME'], id: "home", :nfs => true, :mount_options => ['nolock,vers=3,udp']
+      end
+
+      if File.exist?(CLOUD_CONFIG_PATH)
+        config.vm.provision :file, :source => "#{CLOUD_CONFIG_PATH}", :destination => "/tmp/vagrantfile-user-data"
+        config.vm.provision :shell, :inline => "mv /tmp/vagrantfile-user-data /var/lib/coreos-vagrant/", :privileged => true
+      end
+
+      config.vm.provision :shell, :path => "provision/install_go.sh", :privileged => false
```
Comment (jonboulle) [inline functional/Vagrantfile]: Can't we actually just do this in the cloudinit?

File: functional/config.rb (hunk 23 above)
Comment (kayrus) [inline functional/config.rb]: It is much easier to sync modified Vagrantfile with the original one. So let's use config.rb

File: functional/platform/nspawn.go (hunk 2 above)
Comment (kayrus) [inline functional/platform/nspawn.go]: Sometimes when your environment has problems, this output message will help you to debug.

File: functional/test-docker (hunk 24 above)
Comment (kayrus) [inline functional/test-docker]: If we already have `build-docker` we have to put `test-docker`.

File: functional/Vagrantfile (hunk 25 above)
Comment (kayrus) [inline functional/Vagrantfile]: Great idea. Our Vagrantfile will be identical to original one. Will do that.

File: functional/config.rb (hunk 23 above)
Comment (kayrus) [inline functional/config.rb]: No, we can not. Because in this case we will have to create a hook which will wait until provisioning has been finished. And we don't control cloudinit execution in vagrant. This will complicate things.

File: functional/Vagrantfile (hunk 25 above)
Comment (kayrus) [inline functional/Vagrantfile]: No, we can not. Because in this case we will have to create a hook which will wait until provisioning has been finished. And we don't control cloudinit execution in vagrant. This will complicate things.

File: functional/test [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+export VERSION=$(git describe --dirty)
+export GOROOT=${HOME}/go
+export PATH=${HOME}/go/bin:${PATH}
+
+if [ ! -S "$SSH_AUTH_SOCK" ]; then
+  eval $(ssh-agent)
+fi
+
+chmod 600 functional/fixtures/id_rsa
```
Comment (tixxdz) [inline functional/test]: Please document why we need this one.

File: functional/test-env [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+      echo "$1=$2"
+    fi
+  else
+    echo "could not determine Go version"
+    exit 1
+  fi
+}
+
+export GOBIN=${PWD}/bin
+export GOPATH=${PWD}/gopath
```
Comment (tixxdz) [inline functional/test-env]: I would say if some of these environment are shared between the build and test, then just make a single copy in the root directory of fleet, then source it from there. Not sure if currently it's the case but it would be better if we start the functional tests from the root directory too then cd into functional ? at least one would be able to start them from root directory.

Thank you!

File: functional/user-data [hunk 28]
```diff
@@ -0,0 +1,12 @@
...
+---
+coreos:
+  etcd2:
+    advertise-client-urls: http://$private_ipv4:2379
+    listen-client-urls: http://0.0.0.0:2379,http://0.0.0.0:4001
+  units:
+  - name: etcd2.service
+    command: start
+  update:
+    reboot-strategy: off
```
Comment (tixxdz) [inline functional/user-data]: Adding documentation to this file would also help, why it's here...

Thanks!

File: functional/test-docker [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+echo -e '#!/bin/sh\n LD_LIBRARY_PATH=/opt/lib64 /opt/lib64/ld-linux-x86-64.so.2 /opt/sbin/brctl \$@' > /opt/brctl
+echo -e '#!/bin/sh\n LD_LIBRARY_PATH=/opt/lib64 /opt/lib64/ld-linux-x86-64.so.2 /opt/bin/systemd-nspawn \$@' > /opt/systemd-nspawn
+echo -e '#!/bin/sh\n LD_LIBRARY_PATH=/opt/lib64 /opt/lib64/ld-linux-x86-64.so.2 /opt/bin/systemctl \$@' > /opt/systemctl
+echo -e '#!/bin/sh\n LD_LIBRARY_PATH=/opt/lib64 /opt/lib64/ld-linux-x86-64.so.2 /opt/bin/machinectl \$@' > /opt/machinectl
+chmod +x /opt/brctl /opt/systemd-nspawn /opt/systemctl /opt/machinectl
+source functional/test-env
+eval \$(go env)
+export PATH=\"\${PATH}:/opt:/opt/bin:/opt/sbin\"
+go test github.com/coreos/fleet/functional -ldflags \"\${GLDFLAGS}\" -v
+"
```
Comment (tixxdz) [inline functional/test-docker]: Ok it would be better if we separate this logic into its own, it's too long! and if something is shared then make a single copy of it into the root directory. thanks!

File: functional/platform/nspawn.go (hunk 11 above)
Comment (tixxdz) [inline functional/platform/nspawn.go]: I would suggest to separate these requirements on their own, not hard code them inside the code, just a separate function that does the setup for us in its own file ? @kayrus you could open a PR later for this one perhaps ?

File: functional/test-env (hunk 27 above)
Comment (kayrus) [inline functional/test-env]: Moved test-env into /build-env

File: functional/test (hunk 26 above)
Comment (kayrus) [inline functional/test]: Done

File: functional/user-data (hunk 28 above)
Comment (kayrus) [inline functional/user-data]: Done.

File: functional/platform/nspawn.go [hunk 30]
```diff
@@ -211,10 +211,10 @@ func (nc *nspawnCluster) prepCluster() (err error) {
 		return
 	}
 
-	if !strings.Contains(stdout, "172.17.0.1/16") {
-		_, _, err = run("ip addr add 172.17.0.1/16 dev fleet0")
+	if !strings.Contains(stdout, "172.18.0.1/16") {
```
Comment (jonboulle) [inline functional/platform/nspawn.go]: context for this change?

Comment (kayrus) [inline functional/platform/nspawn.go]: If you have running docker daemon with the default configuration in CoreOS (`172.17.42.1`), these networks intersect and tests don't work.

Comment (jonboulle) [inline functional/platform/nspawn.go]: OK, comment to that effect would help for posterity.

File: functional/Vagrantfile (hunk 25 above)
Comment (jonboulle) [inline functional/Vagrantfile]: Why do you need to wait for provisioning to finish? Can't you just write a one-shot that invokes the equivalent of install_go.sh ?

Comment (jonboulle) [inline functional/Vagrantfile]: Then we can even get rid of that file too

File: functional/README.md [hunk 31]
```diff
@@ -2,52 +2,98 @@
...
-export FLEETCTL_BIN="$(pwd)/bin/fleetctl"
-sudo -E env PATH=$PATH go test github.com/coreos/fleet/functional -v
+
+#### Install VirtualBox
+
+```sh
+source /etc/os-release
+for id in $ID_LIKE $ID; do break; done
+OS_ID=${id:-rhel}
+curl http://download.virtualbox.org/virtualbox/rpm/$OS_ID/virtualbox.repo | sudo tee /etc/yum.repos.d/virtualbox.repo
```
Comment (jonboulle) [inline functional/README.md]: sudo curl -s -O /etc/....

File: functional/README.md [hunk 32]
```diff
@@ -2,52 +2,98 @@
...
+### Centos/Fedora
 
-An example test session using coreos-vagrant follows. This assumes that go is available in `/home/core/go` and the fleet repository in `/home/core/fleet` on the target machine (the easiest way to achieve this is to use shared folders).
+**NOTE**: NFS and Vagrant doesn't work out of the box on CentOS 6.x
+
+#### Install Vagrant
+
+```sh
+sudo yum install -y git nfs-utils
+sudo reboot
```
Comment (jonboulle) [inline functional/README.md]: why is this necessary? o_O

File: functional/Vagrantfile (hunk 25 above)
Comment (reneengelhard) [inline functional/Vagrantfile]: you mean installing go in the test sceript iself? yeah, we could do and actually that was the case before.

But it's cleaner to seperate the steps. Create the VM (and set it up) and actually running the tests. And installking go belongs to the first part so it was actually moved out of the test to a separate provisioner shell script.

File: functional/README.md [hunk 33]
```diff
@@ -2,52 +2,98 @@
...
+wget -q https://www.virtualbox.org/download/oracle_vbox.asc -O- | sudo apt-key add -
+sudo apt-get update
+sudo apt-get install -y build-essential dkms
+sudo apt-get install -y VirtualBox-5.0
+#Previous VirtualBox (if you have problems with nested virtualization, more info here: https://www.virtualbox.org/ticket/14965)
+#sudo apt-get install -y VirtualBox-4.3
 ```
 
-Since the tests utilize `systemd-nspawn`, this needs to be invoked as sudo/root.
+### Centos/Fedora
```
Comment (jonboulle) [inline functional/README.md]: CentOS

File: functional/README.md [hunk 34]
```diff
@@ -2,52 +2,98 @@
...
 
-An example test session using coreos-vagrant follows. This assumes that go is available in `/home/core/go` and the fleet repository in `/home/core/fleet` on the target machine (the easiest way to achieve this is to use shared folders).
+**NOTE**: NFS and Vagrant doesn't work out of the box on CentOS 6.x
```
Comment (jonboulle) [inline functional/README.md]: hence, CentOS >7.x must be used, or something

File: functional/Vagrantfile (hunk 25 above)
Comment (jonboulle) [inline functional/Vagrantfile]: No, I don't mean in the test script. I mean in the [user-data](https://github.com/coreos/fleet/pull/1430/files#diff-832e3e2830d9fbf1ffb2df429089afa0R1)

File: functional/config.rb (hunk 23 above)
Comment (jonboulle) [inline functional/config.rb]: What do we really need from here? Just the shared folder?

Comment (kayrus) [inline functional/config.rb]: channel should be stable, 1 cpu, 512 memory.

File: functional/platform/nspawn.go (hunk 30 above)
Comment (kayrus) [inline functional/platform/nspawn.go]: Already added this info into README.md

File: functional/README.md (hunk 32 above)
Comment (kayrus) [inline functional/README.md]: Again, didn't find time to figure out why NFS can not be started. After reboot it works...

File: functional/README.md (hunk 34 above)
Comment (kayrus) [inline functional/README.md]: do I need to add this text into README.md?

File: functional/README.md (hunk 31 above)
Comment (kayrus) [inline functional/README.md]: -O doesn't work on centos

File: functional/README.md (hunk 33 above)
Comment (kayrus) [inline functional/README.md]: fixed

File: functional/README.md (hunk 32 above)
Comment (jonboulle) [inline functional/README.md]: Just remove this line.

File: functional/README.md [hunk 35]
```diff
@@ -2,52 +2,98 @@
...
+sudo apt-get install -y VirtualBox-5.0
+#Previous VirtualBox (if you have problems with nested virtualization, more info here: https://www.virtualbox.org/ticket/14965)
+#sudo apt-get install -y VirtualBox-4.3
 ```
 
-Since the tests utilize `systemd-nspawn`, this needs to be invoked as sudo/root.
+### CentOS/Fedora
 
-An example test session using coreos-vagrant follows. This assumes that go is available in `/home/core/go` and the fleet repository in `/home/core/fleet` on the target machine (the easiest way to achieve this is to use shared folders).
+**NOTE**: NFS and Vagrant doesn't work out of the box on CentOS 6.x
```
Comment (jonboulle) [inline functional/README.md]: Just change this to "CentOS >= 7.x  is required" or something

File: functional/Vagrantfile (hunk 25 above)
Comment (jonboulle) [inline functional/Vagrantfile]: Ping?

File: functional/config.rb (hunk 23 above)
Comment (jonboulle) [inline functional/config.rb]: I am really struggling to agree that it's simpler to maintain the Vagrantfile in sync with that from coreos-vagrant - [_but not quite_](https://github.com/coreos/fleet/pull/1430/files#diff-a508e808c0fca80a337e9a8d22945e85R142) - rather than just port it to a much simpler Vagrantfile that fulfills our very particular purposes for testing here. 

In case it's not clear why I am pushing back on all these things: every file we add increases the complexity and burden of understanding and maintaining the codebase. I am asking lots of dumb questions here because even though I'm familiar with fleet and CoreOS it's still far from obvious to me why we really need all these parts, and e.g. where I should start making changes if I want to modify things. How can we make this easier for others approaching who have even less context?

**Comment by jonboulle** [comment]
For the sake of moving forward:
i) squash commit history into something sensible
ii) land this
iii) please follow up the comments I've just added

File: functional/README.md (hunk 35 above)
Comment (kayrus) [inline functional/README.md]: it is not really required. someone can definitely configure this.

File: functional/Vagrantfile (hunk 25 above)
Comment (kayrus) [inline functional/Vagrantfile]: @jonboulle we can write oneshot unit file but vagrant will not wait for its result. `run-in-vagrant` has `vagrant up` and `vagrant ssh core-01 -c "sudo ~/fleet/functional/test"`. At the moment last command will be executed once vagrant up provisioning has been finished. If we will use systemd unit - tests will be started before go binary are unpacked.

File: functional/README.md (hunk 35 above)
Comment (kayrus) [inline functional/README.md]: Will add `**NOTE**: NFS and Vagrant doesn't work out of the box on CentOS 6.x, so it is recommended to use CentOS 7.x`

File: functional/README.md (hunk 32 above)
Comment (kayrus) [inline functional/README.md]: Ok

File: functional/config.rb (hunk 23 above)
Comment (kayrus) [inline functional/config.rb]: Well. Let's assume we will add libvirt support for the original Vagrantfile. If we make all modifications (channel, memory, additional cases) it will be harder to merge new changes into this Vagrantfile.

Additional `config.vm.provision` string is an exception. Because without it we have to add more logic into our scripts (i.e. create lock file which will tell `sudo ./test` script that go binaries are downloading and wait until this lock file will be removed, then continue with the tests). If you are ok with these changes - I will do that.
